README.md
# Android.box



### 下载地址：[[fir.im]](https://fir.im/fkw1) [[Github release]](https://github.com/zhihaofans/Android.Box/releases/latest)

### 现有功能：[七大功能](README_features.md)

### 更新记录：[![GitHub release](https://img.shields.io/github/release/zhihaofans/Android.box.svg)![GitHub Release Date](https://img.shields.io/github/release-date/zhihaofans/Android.box.svg)](changelog.md)


changelog.md
# 更新记录

## 更新记录以[Release 列表](https://github.com/zhihaofans/Android.Box/releases)为准

### 0.0.36 (Build 36) - 2018-07-12 13:08:22  7.47 MB

1.哔哩哔哩新增视频封面下载功能

2.新增未完善的通知功能

### 0.0.35 (Build 35) - 2018-07-09 23:51:43  7.52 MB

1.在应用崩溃时自动弹出崩溃信息

2.可以分享Bilibili链接到应用里

3.优化代码

  
### 0.0.34 (Build 34) - 2018-07-06 13:26:49  7.48 MB

1.更换应用崩溃显示库，并在设置里添加查看应用崩溃记录的功能

2.优化数据保存代码，减少应用崩溃的可能性

3.修复设置里"使用内置图片浏览器打开图片链接"无效的BUG

4.删除不必要的代码
  
### 0.0.33 (Build 33) - 2018-07-04 21:29:59 6.70 MB

1.新增应用崩溃信息显示功能

2.新闻盒子当选择的站点只有一个频道时自动选择

3.修复打开非法网址时崩溃的BUG

### 0.0.32 (Build 32) - 2018-06-18 14:07:52 6.68 MB

1.修复App Shortcuts导致闪退的BUG

### 0.0.31 (Build 31) - 2018-06-18 13:13:25 6.68 MB

新闻盒子更新：

1. "RSSHub"站点的“知乎日报”频道由于不稳定移除，新增"V2EX-最新主题"频道

### 0.0.30 (Build 30) - 2018-06-18 00:44:04 6.68 MB

新闻盒子

1. 新增“观点”站点及其频道“观点-首页”

2. 新增“RSSHub”站点及其频道"我以为我很浮夸的微博"、"知乎日报"、“豆瓣正在上映的电影”

### 0.0.29 (Build 29) - 2018-06-16 13:12:24 6.68 MB

1. 新闻盒子新增站点“全球聚合”及其频道“全球聚合-全部”

2. 更新Kotlin库

3. 优化代码

### 0.0.28 (Build 28) - 2018-06-15 01:11:27 6.58 MB

1. 新增手动更新功能

2. 移除指纹验证功能

3. 优化安装包大小

### 0.0.27 (Build 27) - 2018-06-14 00:32:54 6.59 MB

1. 加入“图片链接用内置图片浏览器打开”设置

2. 删除新闻盒子里“数字尾巴”网站的“数字尾巴社区”频道

3. 新闻盒子新增“Pingwest品玩”网站及其频道“品玩转发推荐”

### 0.0.26 (Build 26) - 2018-06-11 14:44:57 6.59 MB

1. 二维码功能重做，内置二维码插件，不再需要额外安装二维码插件，支持扫描和生成二维码，也支持通过分享文本来生成二维码

2. 添加App Shortcuts功能，支持扫/生成二维码、新闻盒子、Server酱

3. 支持选择是否通过Chrome Custom Tabs打开网页

4. 优化代码

### 0.0.25 (Build 25) - 2018-06-03 20:33:18 6.26 MB

1. 新闻盒子的“干货集中营”站点新增“福利”与“Android”频道

2. 新增图片浏览功能，新闻盒子的图片链接会直接用内置功能打开，而不是外部浏览器

3. 新闻盒子右上角重复菜单移除

4. 优化代码

### 0.0.24 (Build 24) - 2018-05-30 06:23:41 3.91 MB

新闻盒子更新了以下内容：

1. 右上角新增刷新按钮

2. 从原来的多个站点改为一个站点对应多个频道

3. 悬浮菜单图标改为白色

4. 优化了新闻盒子的代码

### 0.0.23 (Build 23) - 2018-05-24 01:42:44 3.90 MB

1.更新系统分享菜单里"分享至Server酱选项"的logo

### 0.0.22 (Build 22) - 2018-05-24 01:39:31 3.86 MB

1. 新闻盒子顶栏菜单合并至底部悬浮菜单

2. 新增Server酱服务

### 0.0.20 (Build 20) - 2018-05-12 00:41:32 3.86 MB

1. 删除旧版本的残余代码

2. 新增“分享到剪切板”功能

3. 新闻盒子不再保存上次页码

4. 修复了非首次进入新闻盒子时页码显示为0的BUG

### 0.0.19 (Build 19) - 2018-05-11 14:54:26 3.87 MB

1. 优化移除Fir.im更新检测后的残余代码

### 0.0.18 (Build 18) - 2018-05-11 13:59:03 3.86 MB

1. 引入Bugly用于崩溃检测与应用更新(替代原先简陋的更新方式)

### 0.0.17 (Build 17) - 2018-05-02 00:12:33 3.53 MB

1. 支持Chrome Custom Tabs

### 0.0.16 (Build 16) - 2018-05-01 23:40:27 3.53 MB

1. 修复Bilibili视频弹幕查用户uid功能的线程BUG

### 0.0.15 (Build 15) - 2018-05-01 23:28:48 3.53 MB

1. 更新引用库版本

2. 完善Bilibili视频弹幕查用户uid功能

### 0.0.14 (Build 14) - 2018-04-24 00:03:58 3.51 MB

1. 修复部分库未初始化导致闪退的BUG

### 0.0.13 (Build 13) - 2018-04-23 23:26:56 3.51 MB

1. 修复更新后不会删除更新安装包的BUG

2. 加入了手动删除所有下载文件的选项(主界面右上角菜单->删除临时文件)

### 0.0.12 (Build 12) - 2018-04-14 01:52:42 3.86 MB

1. 修复Bilibili通过弹幕列表的用户hash查询用户个人资料功能开发时残留的文本

### 0.0.11 (Build 11) - 2018-04-14 01:44:53 3.86 MB

1. 增加Bilibili通过弹幕列表的用户hash查询用户个人资料功能

2. 修复当本地版本比服务器高时还是提示更新的BUG

### 0.0.10 (Build 10) - 2018-03-29 22:10:04 3.84 MB

1. 修复检测更新功能BUG

### 0.0.9 (Build 9) - 2018-03-29 21:45:47 3.84 MB

1. 加入检测更新功能

### 0.0.8 (Build 8) - 2018-03-29 19:18:54 3.83 MB

1. 添加天气页面

2. 优化代码

### 0.0.7 (Build 7) - 2018-03-14 00:16:00  6.19 MB

1.新闻盒子添加了"Gank.io"

### 0.0.1-0.0.6

不见了


app/src/main/java/com/zhihaofans/androidbox/mod/SitesMod.kt
package com.zhihaofans.androidbox.mod

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.orhanobut.logger.Logger
import com.zhihaofans.androidbox.data.*
import com.zhihaofans.androidbox.gson.FirimApiLatestUpdate
import com.zhihaofans.androidbox.gson.FirimApiLatestUpdateError
import com.zhihaofans.androidbox.gson.GithubReleaseItem
import com.zhihaofans.androidbox.gson.RsshubFirimGson
import com.zhihaofans.androidbox.kotlinEx.remove
import com.zhihaofans.androidbox.util.*
import okhttp3.CacheControl
import okhttp3.OkHttpClient
import okhttp3.Request
import org.jsoup.Jsoup
import java.io.IOException

/**

 * @author: zhihaofans

 * @date: 2018-12-10 19:09

 */
class NewsSitesMod {
    companion object {
        private val g = Gson()
        private val defaultAppResult = AppInfoResult(false, "", -1, null)
        fun githubReleaseMod(author: String, project: String): AppInfoResult {
            var result = defaultAppResult
            if (author.isEmpty() && project.isEmpty()) {
                result.message = "错误，author或project为空"
            } else {
                val apiUrl = UrlMod.APP_GITHUB_RELEASE.replace("@author@", author).replace("@project@", project)
                val client = OkHttpClient()
                val requestBuilder = Request.Builder().get().cacheControl(CacheControl.Builder().noCache().build()).url(apiUrl)
                val request = requestBuilder.build()
                val call = client.newCall(request)
                try {
                    val response = call.execute()
                    if (response.body() == null) {
                        result.message = ""
                    } else {
                        val jsonData = response.body()!!.string()
                        result = githubApiReleaseJson2Class(author, project, jsonData)
                    }
                } catch (e: IOException) {
                    e.printStackTrace()

                }
            }
            return result
        }

        fun githubApiReleaseJson2Class(author: String, project: String, jsonData: String): AppInfoResult {
            val result = defaultAppResult
            val type = object : TypeToken<List<GithubReleaseItem>>() {}.type
            val github: List<GithubReleaseItem> = g.fromJson(jsonData, type)
            if (github.isEmpty()) {
                result.message = "错误：github.isEmpty"
            } else {
                val lastRelease = github[0]
                Logger.d(lastRelease)
                result.result = AppInfo(author, project, project, "GITHUB_RELEASES", author,
                        if (lastRelease.name.isNullOrEmpty()) lastRelease.tag_name else lastRelease.name,
                        DatetimeUtil.githubUtc2Local(lastRelease.published_at), null,
                        lastRelease.html_url, lastRelease.assets.map {
                    FileList(
                            it.name,
                            it.browser_download_url,
                            it.download_count.toString(),
                            it.updated_at,
                            ConvertUtil.fileSizeInt2string(it.size)
                    )
                }.toMutableList()
                )
                result.success = true
                result.code = 0
            }
            return result
        }

        fun CoolapkWeb(packageName: String): AppInfoResult {
            val webUrl = "https://www.coolapk.com/apk/$packageName"
            val doc = Jsoup.connect(webUrl).get()
            val jsoupUtil = JsoupUtil(doc)
            val result = defaultAppResult
            Logger.d(doc.html())
            val title = jsoupUtil.title()
            if (title != "出错了") {
                val body = jsoupUtil.html("body")
                val a = jsoupUtil.html("p.detail_app_title")
                val b = body.indexOf("window.location.href = \"") + 24
                val c = jsoupUtil.html("p.apk_topba_message").split(" / ")
                val appInfos = jsoupUtil.html("div.apk_left_title > p.apk_left_title_info", 1).split("<br>")
                //Logger.d("appInfos:size:${appInfos.size}\n$appInfos")
                val appName = a.substring(0, a.indexOf("<span class=\"list_app_info\">"))
                val appVersion = jsoupUtil.text("p.detail_app_title > span.list_app_info")
                val appSize = c[0]
                val downloadUrl = body.substring(b, body.indexOf("\"", b))
                //val author = if (appInfos.size != 4) "" else appInfos[3].split("：")[1]
                val rawTime = if (appInfos.size != 4) "" else appInfos[appInfos.size - 3].split("：")[1]
                var newTime = if (appInfos.size != 4) "" else rawTime.replace("-", "/")
                if (newTime.endsWith("天前")) newTime = SystemUtil.datePlus(SystemUtil.nowDate(), -(newTime.substring(0, newTime.length - 2).toIntOrNull()
                        ?: 0))
                Logger.d("rawTime:$rawTime\nnewTime:$newTime")
                val updateTime = if (newTime.isEmpty()) rawTime else newTime
                val downCount = c[1]
                result.result = AppInfo(packageName, null, appName, "COOLAPK_WEB", "", appVersion, updateTime, packageName, webUrl,
                        mutableListOf(FileList(appVersion, downloadUrl, downCount, updateTime, appSize))
                )
                result.success = true
                result.code = 0
            } else {
                result.message = "错误，找不到应用，服务器返回信息：$title"
            }
            return result
        }

        fun FirimApi(pageageName: String, apiToken: String): AppInfoResult {
            val apiUrl = "https://api.fir.im/apps/latest/$pageageName?type=android&api_token=$apiToken"
            Logger.d("apiUrl:$apiUrl")
            val client = OkHttpClient()
            val request = Request.Builder().get().cacheControl(CacheControl.Builder().noCache().build()).url(apiUrl).build()
            val call = client.newCall(request)
            val result = defaultAppResult
            if (pageageName.isEmpty() || apiToken.isEmpty()) {
                result.code = 1
                result.message = "Error: App id or Api token is empty"
                return result
            }
            try {
                val response = call.execute()
                if (response.body() == null) {
                    result.code = 1
                    result.message = "Error: response.body is null"
                } else {
                    val jsonData = response.body()!!.string()
                    if (jsonData.isEmpty()) {
                        result.code = 1
                        result.message = "Error: jsonData is null"

                    } else {
                        result.message = "Error:Gson to firimV1"
                        val firimApiLatestUpdateError = g.fromJson(jsonData, FirimApiLatestUpdateError::class.java)
                        result.message = ""
                        if (firimApiLatestUpdateError.code != null && firimApiLatestUpdateError.errors != null) {
                            result.code = 2
                            result.message = "Error! Code:${firimApiLatestUpdateError.code}"
                            val errors = firimApiLatestUpdateError.errors.exception
                            var errorIndex = 0
                            errors.map {
                                result.message += when (errorIndex) {
                                    0 -> ", message:$it"
                                    else -> "|$it"
                                }
                                errorIndex++
                                it
                            }
                        } else {
                            result.message = "Error:Gson to firimV1"
                            val latestUpdate = g.fromJson(jsonData, FirimApiLatestUpdate::class.java)
                            result.message = ""
                            Logger.d(latestUpdate.updated_at)
                            val updateTime = DatetimeUtil.unixTime2date("${latestUpdate.updated_at}000".toLong())
                            result.result = AppInfo(pageageName, apiToken, latestUpdate.name, "FIRIM_V1", "",
                                    latestUpdate.versionShort + "(${latestUpdate.version})",
                                    updateTime, pageageName, latestUpdate.update_url,
                                    mutableListOf(FileList(latestUpdate.versionShort + "(${latestUpdate.version})", latestUpdate.install_url,
                                            null, updateTime, ConvertUtil.fileSizeInt2string(latestUpdate.binary.fsize)))
                            )
                            result.success = true
                            result.code = 0
                            result.message = "Success"

                        }

                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
                result.code = -1
                if (result.message.isNotEmpty()) {
                    result.message = "Error:IOException"
                }
            } finally {
                return result
            }
        }

        fun FirimRsshub(projectId: String): AppInfoResult {
            val apiUrl = "https://rsshub.app/fir/update/$projectId.json"
            Logger.d("apiUrl:$apiUrl")
            val client = OkHttpClient()
            val request = Request.Builder().get().cacheControl(CacheControl.Builder().noCache().build()).url(apiUrl).build()
            val call = client.newCall(request)
            val result = defaultAppResult
            if (projectId.isEmpty()) {
                result.code = 1
                result.message = "Error: Project id or Api token is empty"
                return result
            }
            try {
                val response = call.execute()
                if (response.body() == null) {
                    result.code = 1
                    result.message = "Error: response.body is null"
                } else {
                    val jsonData = response.body()!!.string()
                    if (jsonData.isEmpty()) {
                        result.code = 1
                        result.message = "Error: jsonData is null"
                    } else {
                        result.message = "Error:Gson to firimV2"
                        val rsshubFirimGsons = g.fromJson(jsonData, RsshubFirimGson::class.java)
                        if (rsshubFirimGsons.title == "RSSHub" || rsshubFirimGsons.home_page_url == "https://github.com/DIYgod/RSSHub") {
                            result.message = "Error:Title is RSSHub"
                        } else if (rsshubFirimGsons.items.size == 0) {
                            result.message = "Error:Not item"
                        } else {
                            result.message = ""
                            val webUrl = rsshubFirimGsons.home_page_url
                            val updateItem = rsshubFirimGsons.items[0]
                            val updateTime = updateItem.date_published.remove("&#34;")
                                    .replace("-", "/")
                                    .replace("T", " ")
                                    .replace(".000Z", "")
                            val updateVersion = updateItem.title
                            result.result = AppInfo(projectId, null, rsshubFirimGsons.title, "FIRIM_V2", "",
                                    updateVersion, updateTime, null, webUrl, mutableListOf())
                            result.success = true
                            result.code = 0
                            result.message = "Success"
                        }
                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
                result.code = -1
                if (result.message.isNotEmpty()) {
                    result.message = "Error:IOException"
                }
            } finally {
                return result
            }
        }

        fun WandoujiaWeb(packageName: String): AppInfoResult {
            val webUrl = "https://www.wandoujia.com/apps/$packageName"
            val doc = Jsoup.connect(webUrl).get()
            val jsoupUtil = JsoupUtil(doc)
            val result = defaultAppResult
            Logger.d(doc.html())
            val title = jsoupUtil.title()
            val body = jsoupUtil.body()
            if (body == null) {
                result.message = "错误，找不到应用，服务器返回信息"
            } else {
                val webCode = body.attr("param-f") ?: "404"
                if (webCode.isEmpty() || webCode != "detail" || title == "豌豆荚") {
                    result.message = "错误，找不到应用，服务器返回信息"
                } else {
                    val appName = body.attr("data-title") ?: packageName
                    val appVersion = jsoupUtil.html("dl.infos-list > dd", 3).replace("&nbsp;", "")
                    val appSize = jsoupUtil.attr("dl.infos-list > dd > meta", "content")
                    val downloadUrl = jsoupUtil.link("a.install-btn")
                    val author = jsoupUtil.html("span.dev-sites")
                    val updateTime = jsoupUtil.attr("time#baidu_time", "datetime")
                    val downCount = jsoupUtil.html("span.item.install > i")
                    result.result = AppInfo(packageName, null, appName, "WANDOUJIA_V1", author, appVersion, updateTime, packageName, webUrl,
                            mutableListOf(FileList(appVersion, downloadUrl, downCount, updateTime, appSize))
                    )
                    result.success = true
                    result.code = 0
                }
            }
            return result
        }

        fun MyAppWeb(packageName: String): AppInfoResult {
            val webUrl = "http://android.myapp.com/myapp/detail.htm?apkName=$packageName"
            val result = defaultAppResult
            if (packageName.isEmpty()) return result
            val doc = Jsoup.connect(webUrl).get()
            val jsoupUtil = JsoupUtil(doc)
            Logger.d(doc.html())
            val body = jsoupUtil.body()
            if (body == null) {
                result.message = "错误，找不到应用，服务器返回空白信息"
            } else {
                val emptyElement = jsoupUtil.textorNull("search-none-text")
                if (emptyElement.isNullOrEmpty()) {
                    val appName = jsoupUtil.htmlorNull("div.det-name-int") ?: packageName
                    val appIcon = jsoupUtil.img("div.det-icon >img")
                    Logger.d("appIcon:$appIcon")
                    val appSize = jsoupUtil.html("div.det-size")
                    val downloadUrl = jsoupUtil.attr("a.det-down-btn", "data-apkurl")
                    Logger.d("downloadUrl:$downloadUrl")
                    val appVersion = jsoupUtil.html("div.det-othinfo-data", 0)
                    val updateUnixTime = jsoupUtil.attr("#J_ApkPublishTime", "data-apkpublishtime").toLong()
                    val updateTime = DatetimeUtil.unixTime2date(updateUnixTime * 1000)
                    Logger.d("updateUnixTime:$updateUnixTime")
                    Logger.d("updateTime:$updateTime")
                    val author = jsoupUtil.html("div.det-othinfo-data", 2)
                    val downCount = jsoupUtil.html("div.det-ins-num")
                    result.result = AppInfo(packageName, null, appName, "MYAPP", author, appVersion, updateTime, packageName, webUrl,
                            mutableListOf(FileList(appVersion, downloadUrl, downCount, updateTime, appSize))
                    )
                    result.success = true
                    result.code = 0
                } else {
                    result.message = "错误，找不到应用，服务器返回信息($emptyElement)"
                }
            }
            return result
        }

        /*fun ApkPure(packageName: String): AppInfoResult {
            val webUrl = "https://apkpure.com/cn/$packageName"
            val result = defaultAppResult
            if (packageName.isEmpty()) return result
            val doc = Jsoup.connect(webUrl).get()
            val jsoupUtil = JsoupUtil(doc)
            Logger.d(doc.html())
            val body = jsoupUtil.body()
            if (body == null) {
                result.message = "错误，找不到应用，服务器返回空白信息"
            } else {
                val webTitle = jsoupUtil.titleOrNull()
                if (webTitle.isNullOrEmpty() || webTitle == "404") {
                    result.message = "错误，找不到应用"
                } else {
                    val appInfoElement = jsoupUtil.findInElementsText("script[type=\"text/javascript\"]", "var comment_config = {")
                    if (appInfoElement.size != 1) {
                        result.message = "错误，获取应用信息失败 (appInfoElement.size != 1)"
                    } else {
                        val appInfoStr = appInfoElement[0].text()
                        if (appInfoStr.isEmpty()) {
                            result.message = "错误，获取应用信息失败 (appInfoStr.isEmpty)"
                        } else {
                            var appInfoStartFrom = appInfoStr.find("var comment_config = {")
                            if (appInfoStartFrom == -1) {
                                result.message = "错误，获取应用信息失败 (appInfoStartFrom.size == 1)"
                            } else {
                                appInfoStartFrom += 22
                                val appName = jsoupUtil.textorNull("div.title-like > h1")
                                        ?: packageName
                                val appIcon = jsoupUtil.img("div.icon > img")
                                val appSize = jsoupUtil.text(".ny-down > a > span.fsize > span")
                                val author = jsoupUtil.html("ul.version-ul > li > p", 1)
                                val appVersion = jsoupUtil.html("ul.version-ul > li > p", 3)
                                val updateTime = jsoupUtil.html("ul.version-ul > li > p", 5)
                                val mobileUrl = "https://m.apkpure.com/store/apps/details?id=$packageName"
                                result.result = AppInfo(packageName, null, appName, "APKPURE",
                                        author, appVersion, updateTime, packageName, mobileUrl, mutableListOf()
                                )
                                result.success = true
                                result.code = 0
                            }

                        }
                    }
                }
            }
            return result
        }*/

    }
}

class XXDownSitesMod {
    companion object {
        fun AcfunVideoThumbnail(url: String): XXDownResultData? {
            val call = HttpUtil.httpClientGetCall(url, XXDownSite.headerbuild)
            try {
                val response = call.execute()
                val body = response.body()
                        ?: return XXDownResultData(false, "Http body is null", listOf())
                val bodyStr = body.string()
                if (!response.isSuccessful) return XXDownResultData(false, "Http get error(${response.message()})", listOf())
                if (bodyStr.isEmpty()) return XXDownResultData(false, "Http body is empty", listOf())
                Logger.d(bodyStr)
                val doc = Jsoup.parse(bodyStr)
                val jsoupUtil = JsoupUtil(doc)
                val image = jsoupUtil.attrOrNull("div#pageInfo", "data-pic")
                return when {
                    image == null -> XXDownResultData(false, "image is null", listOf())
                    image.isEmpty() -> XXDownResultData(false, "image is empty", listOf())
                    else -> XXDownResultData(true, "OK", listOf(XXDownResultUrlData(image, XXDownUrlType.image)))
                }
            } catch (e: Exception) {
                e.printStackTrace()
                return XXDownResultData(false, "Http get error(Exception)", listOf())
            }
        }

        fun BilibiliVideoThumbnail(url: String): XXDownResultData? {
            val call = HttpUtil.httpClientGetCall(url, XXDownSite.headerbuild)
            try {
                val response = call.execute()
                val body = response.body()
                        ?: return XXDownResultData(false, "Http body is null", listOf())
                val bodyStr = body.string()
                if (!response.isSuccessful) return XXDownResultData(false, "Http get error(${response.message()})", listOf())
                if (bodyStr.isEmpty()) return XXDownResultData(false, "Http body is empty", listOf())
                Logger.d(bodyStr)
                val doc = Jsoup.parse(bodyStr)
                val jsoupUtil = JsoupUtil(doc)
                Logger.d(jsoupUtil.html("head > meta[itemprop='image']"))
                val image = jsoupUtil.attr("head > meta[itemprop=\"image\"]", "content")
                return when {
                    image.isEmpty() -> XXDownResultData(false, "image is empty", listOf())
                    else -> XXDownResultData(true, "OK", listOf(XXDownResultUrlData(image, XXDownUrlType.image)))
                }
            } catch (e: Exception) {
                e.printStackTrace()
                return XXDownResultData(false, "Http get error(Exception)", listOf())
            }
        }

        fun githubReleaseXX(url: String, allowPre: Boolean = false): XXDownResultData? {
            if (url.startsWith(UrlMod.XXDOWN_SITE_GITHUB_RELEASE)) {
                var mUrl = url.remove(UrlMod.XXDOWN_SITE_GITHUB_RELEASE)
                if (mUrl.endsWith("/")) mUrl = mUrl.substring(0, mUrl.length - 2)
                val mList = mUrl.split("/")
                if (mList.size != 2) return null
                val author = mList[0]
                val project = mList[1]
                val githubReleaseMod = NewsSitesMod.githubReleaseMod(author, project)
                try {
                    if (!githubReleaseMod.success) return XXDownResultData(false, "Get error(${githubReleaseMod.message})", listOf())
                    if (githubReleaseMod.result == null) return XXDownResultData(false, "Result is null", listOf())
                    val fileList = githubReleaseMod.result!!.fileList
                    return when {
                        fileList.isNullOrEmpty() -> XXDownResultData(false, "fileList is null or empty", listOf())
                        else -> XXDownResultData(true, "OK", fileList.map {
                            XXDownResultUrlData(it.url, XXDownUrlType.other)
                        }.toList())
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                    return XXDownResultData(false, "Need Github release url", listOf())
                }
            } else {
                return null
            }
        }
    }
}